architecture/spmc.rst

```diff
@@ -0,0 +1,269 @@
+====
+SPMC
+====
+
+This document describes the SPMC (S-EL1) implementation for OP-TEE.
+More information on the SPMC can be found in the FF-A specification can be
+found in the
+`FF-A spec <https://developer.arm.com/documentation/den0077/latest>`_.
+
+.. toctree::
+ :numbered:
+
+SPMC Responsibilities
+=====================
+
+The SPMC is a critical component in the FF-A flow. Some of its mayor
+responsibilities are:
+
+- Initialisation and run-time management of the SPs:
+	The SPMC component is responsible for initialisation of the
+	Secure Partitions (Loading the Image, Setting up the stack, HEAP, ...).
+- Routing messages between endpoints:
+	The SPMC is responsible for passing FF-A messages from normal world
+	to SPs and back. It also responsible for passing FF-A messages between
+	SPs.
+- Memory management:
+	The SPMC is responsible for the memory management of the SPs. Memory
+	can be shared between SPs and between a SP to the normal world.
+
+This document describes OP-TEE as a S-EL1 SPMC.
+
+Secure Partitions
+=================
+Secure Partitions (SPs) are the endpoints used in the FF-A protocol. When
+OP-TEE is used as a SPMC SPs run primarily inside S-EL0.
+
+The SPMC also exposes the OP-TEE core, privileged mode, as an secure endpoint
+itself. This is used to handle all GlobalPlaform programming mode operations.
+All GlobalPlatform messages are encapsulated inside FF-A messages.
+The OP-TEE endpoint will unpack the messages and afterwards handle them as
+standard OP-TEE calls. This is needed as TF-A (S-EL3) does only allow
+FF-A messages to be passed to the secure world when the SPMD is enabled.
+
+SPs run from the initial boot of the system until power down and don't have any
+built-in session management compared to GPD TEE TAs. The only means of
+communicating with the outside world is through messages defined in the FF-A
+specification. The context of a SP is saved between executions.
+
+The
+`Trusted Service <https://www.trustedfirmware.org/projects/trusted-services/>`_
+repository includes the libsp libary which export all needed functions to build
+a S-EL0 SP. It also includes many examples of how to create and implement a SP.
+
+SPMC Program Flow
+=================
+SP images are stored OP-TEE image as early TAs are: the binary images are
+embedded in OP-TEE Core in a read-only data section.
+This makes it possible to load SPs during boot. Early TSs are embedded inside
+the OP-TEE image. This makes it possible to load SPs during the initialisation
+of OP-TEE and no rich-os is needed in the normal world. Ldelf is used to load
```

Comment:
- You are saying the same thing twice, aren't you?
- ` ``ldelf`` `

Reply:
Done